packages/jsii-pacmak/lib/targets/kotlin/declarationnode.ts

```diff
@@ -0,0 +1,49 @@
+import { Assembly, Type } from 'jsii-reflect';
+
+export class DeclarationNode {
+  public static of(assembly: Assembly): DeclarationNode {
+    const tree = new DeclarationNode('');
+    for (const type of assembly.types) {
+      tree.register(type, DeclarationNode.getTypeName(type).split('.'));
+    }
+    return tree;
+  }
+
+  private static getTypeName(type: Type): string {
+    if (type.namespace) {
+      return `${type.namespace}.${type.name}`;
+    }
+
+    return type.name;
+  }
+
+  private _type?: Type;
+  private _children: { [name: string]: DeclarationNode } = {};
+
+  private constructor(public readonly name: string) {}
+
+  public get type(): Type | undefined {
+    return this._type;
```

Comment:
Feels odd for me to have this being late-bound (but willing to give the benefits of doubt here)

Reply:
As we traverse the declarations, it seems there is no guarantee that `foo` will appear earlier than `foo.bar`. So I firstly register a declaration name alone and bind the type when it's found.